### PR TITLE
ci: improve Docker Release workflow

### PR DESCRIPTION
**Pull request by Juneezee**
### What I did

Setup the Docker Release workflow to build and push image to both Dockerhub and GitHub Container Registry on tag push.

This pull request aims to fix Issue #518.

This is an improvement to PR #504.

### How I did it

1. Renamed `.github/workflows/ghci.io.yml` into `.github/workflows/docker.yml`. This is because the Docker Release workflow now pushes to both **Dockerhub** and **GitHub Container Registry**. @lorenzo You will need to add `DOCKERHUB_USERNAME` and `DOCKERHUB_TOKEN` as the repository secrets in order to push to Dockerhub.

2. Removed both `docker/hooks/build` and `docker/hooks/push`. The Docker Release workflow now uses the official [docker/build-push-action](https://github.com/docker/build-push-action) for building and pushing image. 

3. Updated [upx/upx](https://github.com/upx/upx) from `3.94` to `3.96`.

4. Correctly tag and push `scratch` image. Previously only `debian-distro` and `alpine-distro` was targeted and tagged. This is fixed by removing the `AS distro` from `FROM scratch AS distro`. Now the `scratch`, `alpine`, and `debian` images will be built and pushed correctly by the workflow. The image below shows the size of `scratch`, `alpine`, and `debian` image respectively.
image

5. For example, if the tag pushed is `v2.0.0`, 
    - `scratch` image will be tagged as `latest` and `2.0.0`
    - `alpine` image will be tagged as `latest-alpine` and `2.0.0-alpine`
    - `debian` image will be tagged as `latest-debian` and `2.0.0-debian`

### How to verify it

I have run the Docker Release workflow on my fork (https://github.com/Juneezee/hadolint/actions/runs/670702218).

image

#### Scratch image
image

#### Alpine image
image

#### Debian image
image

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+          platforms: linux/amd64
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
Comment (Juneezee) [inline .github/workflows/docker.yml:30]: @lorenzo  You will need to add DOCKERHUB_USERNAME and DOCKERHUB_TOKEN as the repository secrets in order to push to Dockerhub.

**Comment by Juneezee** [comment]
@lorenzo In addition to #518, I think this is likely because the package visibility has not been set to **public** yet.

This is what I see when I navigate to the `Packages` section of `hadolint` organisation. There is no `hadolint` package listed in there.

image

---

It should show the `hadolint` package if the package visbility is set to public. For example, the image below is one of the organisation that I own and has two public packages:

image

According to [Configuring visibility of container images for an organization](https://docs.github.com/en/packages/guides/configuring-access-control-and-visibility-for-container-images#configuring-visibility-of-container-images-for-an-organization), the default visibility is private unless it is explicitly set to public.

**Comment by lorenzo** [comment]
@Juneezee I just made it public. Thanks for the guidance :)

**Comment by lorenzo** [comment]
@Juneezee Currently, due to what I believe is a bug, we are pushing `latest` as whatever gets merged into master. Do you think we should keep that or just mark `latest` for actual tags?

**Comment by Juneezee** [comment]
> @Juneezee Currently, due to what I believe is a bug, we are pushing `latest` as whatever gets merged into master. Do you think we should keep that or just mark `latest` for actual tags?

This is no longer the case after this pull request is merged. According to line 3 - 7 in `.github/workflows/docker.yml`,

```
name: Docker Release

on:
  # Trigger the workflow on the new 'v*' tag created
  push:
    tags:
      - "v*"
```

This workflow will only run when a tag starting with `v` has been pushed. I have configured the workflow in a way such that the `v` prefix will be remove when tagging the image. For example, if the tag `v2.0.0` is pushed, the workflow will tag the `scratch` image as `hadolint/hadolint:latest` and `hadolint/hadolint:2.0.0`.

**Comment by lorenzo** [comment]
perfect, thanks so much for your help!

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        run: |
+          if [ "${{ matrix.distro }}" = 'scratch' ]; then
+            echo ::set-output name=target::''
+            echo ::set-output name=latest::latest
+            echo ::set-output name=tag::"${GITHUB_REF/refs\/tags\/v/}"
+          else
+            echo ::set-output name=target::"${{ matrix.distro }}-distro"
+            echo ::set-output name=latest::"latest-${{ matrix.distro }}"
+            echo ::set-output name=tag::"${GITHUB_REF/refs\/tags\/v/}-${{ matrix.distro }}"
+          fi
```
Comment (Juneezee) [inline .github/workflows/docker.yml:50]: This is where the workflow will set the target and tag for `docker build` command appropriately.

**Pull request merged**